datafusion/physical-plan/src/aggregates/row_hash.rs

```diff
@@ -266,6 +266,12 @@ pub(crate) struct GroupedHashAggregateStream {
 
     /// The spill state object
     spill_state: SpillState,
+
+    /// Optional soft limit on the number of `group_values` in a batch
+    /// If the number of `group_values` in a single batch exceeds this value,
+    /// the `GroupedHashAggregateStream` operation immediately switches to
+    /// output mode and emits all groups.
+    group_values_soft_limit: Option<usize>,
```

Comment:
👍

Reply:
Thanks. I believe this is ready for a look. Please let me know if you think I should break it down into smaller PRs.

Reply:
I can't see any good way to break this down (other than maybe breaking out the benchmarks), so this is fine